include/rmm/mr/memory_resource.hpp

```diff
@@ -0,0 +1,427 @@
+   * If an implementation does not support streams, it should return memory which
+   * can be used on any stream.
+   *
+   * @throws Nothing.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param stream Stream on which to perform deallocation
+   */
+  void deallocate(void* p, std::size_t bytes, cuda_stream_view stream)
+  {
+    deallocate(p, bytes, default_alignment<kind>::value, stream);
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by \p p.
+   *
+   * `p` must have been returned by a prior call to `allocate(bytes,stream)` on
+   * a `stream_memory_resource` that compares equal to `*this`, and the storage
+   * it points to must not yet have been deallocated, otherwise behavior is
+   * undefined.
+   *
+   * The memory returned must be available for immediate use on given stream.
+   * If an implementation does not support streams, it should return memory which
+   * can be used on any stream.
+   *
+   * @throws Nothing.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param alignment The alignment of the allocation, as was passed to the
+   * call to `allocate` that returned `p`.
+   * @param stream Stream on which to perform deallocation
+   */
+  void deallocate(void* p, std::size_t bytes, std::size_t alignment, cuda_stream_view stream)
+  {
+    do_dealloc_async(p, bytes, alignment, stream);
+  }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return bool true if the resource supports get_mem_info, false otherwise.
+   */
+  virtual bool supports_get_mem_info() const noexcept = 0;
+
+  /**
+   * @brief Queries the amount of free and total memory for the resource.
+   *
+   * @param stream the stream whose memory manager we want to retrieve
+   *
+   * @returns a std::pair<size_t,size_t> which contains free memory in bytes
+   * in .first and total amount of memory in .second
+   */
+  std::pair<std::size_t, std::size_t> get_mem_info(cuda_stream_view stream) const
+  {
+    return do_get_mem_info(stream);
+  }
+
+ private:
+  /**
+   * @brief Allocates memory of size at least \p bytes.
+   *
+   * The returned pointer will be aligned at least to the required alignment.
+   *
+   * The memory returned must be available for immediate use on given stream.
+   * If an implementation does not support streams, it should return memory which
+   * can be used on any stream.
+   *
+   * @param bytes The size of the allocation
+   * @param stream Stream on which to perform allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  virtual void* do_alloc_async(std::size_t bytes, std::size_t alignment, cuda_stream_view stream) = 0;
```

Comment:
We shouldn't have abbreviated and non-abbreviated versions of the same word in the same code. 
```suggestion
  virtual void* do_allocate_async(std::size_t bytes, std::size_t alignment, cuda_stream_view stream) = 0;
```

Reply:
Done.